### PR TITLE
Git info store object 

### PR DESCRIPTION
**Pull request by daschuer**
This adds a git Info store project to speed up building when only the git info is updated. 
This uses also the build date in case of tar ball builds unless a GIT_COMMIT_DATE is passed to cmake along with a GIT _DESCRIBE

### CODE REVIEW

File: res/skins/Shade/skin.xml [hunk 1]
```diff
@@ -1,4 +1,5 @@
   <!--
+
```
Comment (Holzhaus) [inline res/skins/Shade/skin.xml:2]: unrelated change

File: src/util/versionstore.cpp [hunk 2]
```diff
@@ -114,24 +110,24 @@ QString VersionStore::platform() {
...
-    QString gitVersion = VersionStore::gitDescribe();
+    QString gitVersion = GitInfoStore::describe();
     if (gitVersion.isEmpty()) {
         gitVersion = QStringLiteral("unknown");
     }
 
     QString gitBranch = VersionStore::gitBranch();
     if (!gitBranch.isEmpty()) {
-        gitVersion.append(QStringLiteral(" (") + gitBranch + QStringLiteral(" branch)"));
+        gitVersion.append(QStringLiteral(" (") + gitBranch + QChar(')'));
```
Comment (Holzhaus) [inline src/util/versionstore.cpp:130]: I'm okay with this change to make the about window a bit less wide.

**Comment by Holzhaus** [comment]
Some pre-commit issues.

File: src/util/versionstore.cpp (hunk 2 above)
Comment (daschuer) [reply src/util/versionstore.cpp:130]: we can even remove the branch name if it matches the version

**Comment by daschuer** [comment]
Done

File: src/util/versionstore.cpp (hunk 2 above)
Comment (Holzhaus) [reply src/util/versionstore.cpp:130]: We may, but I don't think it's necessary. In any case this should happen in the cpp code, not in the CMake code.

Comment (Holzhaus) [reply src/util/versionstore.cpp:130]: ah you already did.

**Comment by Holzhaus** [comment]
> Done

Still pre-commit issues.

**Comment by daschuer** [comment]
No idea why this passes locally the second time. :-/
Anyway, should work now.

File: src/util/gitinfostore.cpp [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+const char* GitInfoStore::date() {
+    return GIT_COMMIT_DATE;
+};
+
+//static
+int GitInfoStore::commitCount() {
+#ifdef GIT_COMMIT_COUNT
+    return GIT_COMMIT_COUNT;
+#else
+    return 0;
```
Comment (Holzhaus) [inline src/util/gitinfostore.cpp:26]: -1 maybe? or use std::optional?

File: src/util/gitinfostore.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+class GitInfoStore {
+  public:
+    static const char* branch();
+    static const char* describe();
+    static const char* date();
+    int commitCount();
```
Comment (Holzhaus) [inline src/util/gitinfostore.h:8]: static?

File: CMakeLists.txt [hunk 5]
```diff
@@ -1607,16 +1607,27 @@ get_target_property(MIXXX_BUILD_FLAGS mixxx-lib COMPILE_OPTIONS)
...
 configure_file(src/version.h.in src/version.h @ONLY)
 
+if(GIT_COMMIT_DATE AND NOT GIT_COMMIT_DATE MATCHES "^[0-9]*-[0-9]*-[0-9]*T[0-9]*\\:[0-9]*\\:[0-9]*[+-][0-9]*\\:[0-9]*$")
+  message(FATAL_ERROR "GIT_COMMIT_DATE requires strict ISO 8601 format %Y-%m-%dT%H:%M:%SZ")
+endif()
+
 add_custom_target(mixxx-gitinfo
   COMMAND ${CMAKE_COMMAND}
+    -DGIT_DESCRIBE="${GIT_DESCRIBE}"
+    -DGIT_COMMIT_DATE="${GIT_COMMIT_DATE}"
```
Comment (Holzhaus) [inline CMakeLists.txt:1617]: What about CPackConfig? I think you need to copy these variables into the CPack domain as well, and set them before including the `GitInfo` module.

**Comment by Holzhaus** [comment]
There are merge conflicts now.

File: CMakeLists.txt (hunk 5 above)
Comment (Holzhaus) [reply CMakeLists.txt:1617]: @daschuer ping.

Comment (daschuer) [reply CMakeLists.txt:1617]: Oh yes, we should allow to use the override values in cpack as well.

File: src/util/gitinfostore.cpp (hunk 3 above)
Comment (daschuer) [reply src/util/gitinfostore.cpp:26]: 0 fits, because we have 0 commits if we are off the git worktree

**Comment by daschuer** [comment]
Done.

File: CMakeLists.txt [hunk 6]
```diff
@@ -1607,16 +1607,27 @@ get_target_property(MIXXX_BUILD_FLAGS mixxx-lib COMPILE_OPTIONS)
...
     -DINPUT_FILE="${CMAKE_CURRENT_SOURCE_DIR}/src/gitinfo.h.in"
     -DOUTPUT_FILE="${CMAKE_CURRENT_BINARY_DIR}/src/gitinfo.h"
     -P "${CMAKE_CURRENT_SOURCE_DIR}/cmake/scripts/gitinfo.cmake"
   COMMENT "Update git version information in gitinfo.h"
   BYPRODUCTS "${CMAKE_CURRENT_BINARY_DIR}/src/gitinfo.h"
   WORKING_DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
 )
-add_dependencies(mixxx-lib mixxx-gitinfo)
+
+add_library(mixxx-gitinfostore STATIC EXCLUDE_FROM_ALL
```
Comment (Holzhaus) [inline CMakeLists.txt:1626]: What's the reason for making this a separate library? Why not add it to mixxx-lib instead?

Comment (daschuer) [reply CMakeLists.txt:1626]: To improve build time. Re-linking it to mixxx-lib takes long, even if just the git info changes. 
From this point of view it is better to have more lbs than less. This will also reduce the memory requirements on Windows.

**Comment by uklotzde** [comment]
Building a local branch fails now:

```
CMake Error at CMakeLists.txt:1628 (message):
  GIT_COMMIT_DATE requires strict ISO 8601 format %Y-%m-%dT%H:%M:%SZ
```

**Comment by uklotzde** [comment]
Do we have updated build instructions for this use case??

**Comment by uklotzde** [comment]
This is what I have added to my build script:

```
GIT_COMMIT_DESCRIPTION=$(git describe --tags --always --dirty=-modified)
GIT_COMMIT_DATE=$(git show --quiet --format=%cd --date=short)
```

```
GIT_COMMIT_DESCRIPTION=2.4-alpha-5721-g528306e196
GIT_COMMIT_DATE=2021-05-17
```

**Comment by Holzhaus** [comment]
@uklotzde can you run the git commit date command directly and post the output?

**Comment by uklotzde** [comment]
`--date=iso-strict` works

**Comment by Holzhaus** [comment]
Ah, please use `--format*%cI`.

**Comment by Holzhaus** [comment]
Is this for Fedora? If you're building from the git tree setting those variables locally should not be necessary at all.

**Comment by uklotzde** [comment]
The `--date` argument could be omitted.

We need to build from a plain file archive, no Git repo available. I added them explicitly for continuously testing the build. Just to notice early when the build breaks because of invalid assumptions.

**Comment by uklotzde** [comment]
Ok, I copied the new variables and their commands literally.
